iceoryx_posh/include/iceoryx_posh/internal/popo/server_impl.hpp

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_POSH_POPO_SERVER_IMPL_HPP
+#define IOX_POSH_POPO_SERVER_IMPL_HPP
+
+#include "iceoryx_posh/capro/service_description.hpp"
+#include "iceoryx_posh/internal/popo/base_server.hpp"
+#include "iceoryx_posh/internal/popo/request_deleter.hpp"
+#include "iceoryx_posh/internal/popo/response_deleter.hpp"
+#include "iceoryx_posh/internal/popo/rpc_interface.hpp"
+#include "iceoryx_posh/popo/request.hpp"
+#include "iceoryx_posh/popo/response.hpp"
+#include "iceoryx_posh/popo/server_options.hpp"
+#include "iceoryx_posh/popo/trigger_handle.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+
+namespace iox
+{
+namespace popo
+{
+template <typename Req, typename Res, typename BaseServerT = BaseServer<>>
+class ServerImpl : public BaseServerT, public RpcInterface<Response<Res>>
+{
+    static_assert(!std::is_void<Req>::value, "The type `Req` must not be void. Use the UntypedServer for void types.");
```

Comment:
You have here and in `ClientImpl` some code duplication. What about introducing a `HeaderConcept` like:
```cpp
template <typename T>
struct HeaderConcept {
    static_assert(!std::is_void<T>::value, "Must not be void");
    static_assert(!std::is_const<T>::value, "Must not be const");
    static_assert(!std::is_reference<T>::value, "Must not be a reference");
    static_assert(!std::is_pointer<T>::value, "Must not be a pointer");
    using Evaluate = void; // required and has to be used with typename HeaderConcept<int>::Evaluate otherwise the compiler ignores the static_assert's
};
```

And use it like:
```cpp
template <typename Req, typename Res, typename BaseServerT = int,
          typename RequiresForRequest = typename HeaderConcept<Req>::Evaluate,
          typename RequiresForResponse = typename HeaderConcept<Res>::Evaluate>
class ServerImpl {};
```

Then one would see that a header concept is required (looks a bit like C++20), if something changes we have only one piece of code we have to adjust and not 4 different pieces.

Reply:
@elfenpiff Looks interesting. Can be extended with checks for specific members etc. in the concept should this be needed (with SFINAE). I just do not like the need to introduce artificial template args for this (as with other meta-programming e.g. `enable_if`) which are not really supposed to be used apart from the default values.

I think it is possible/better to transfer the aggregated condition value to some internal bool, i.e. basically create a type trait for headers. Then we can `static_assert` on this as before, like `static_assert(HeaderConcept<Res>::value, "concept not satisfied"). I leave out the details, but its basically a `constexpr bool` on your conditions above.

Then we do not have artificial template args.

Reply:
@elfenpiff this might be tricky since the asserts have subtle variations

Reply:
@elfenpiff after a closer look, forget what I wrote

Reply:
@elfenpiff I agree with @MatthiasKillat to not add additional template args so I adjusted your proposal a bit.